Advanced partitioning customizations (COMPOSER-2355)

achilleas-k opened the pull request:
This PR adds a new advanced partitioning customization to the blueprint that has the following structure:
```yaml
{
  "customizations": {
    "partitioning": {
      "plain": {
        "filesystems": [
          {
            "mountpoint": "",
            "minsize": 0,
            "label": "",
            "fstype": ""
          }
        ],
        "lvm": {
          "volume-groups": [ # only one supported, but maybe we can support multiple in the future
            {
              "name": "",
              "logical-volumes": [ # similar to filesystems but with lvm stuff
                {
                  "mountpoint": "",
                  "minsize": 0,
                  "label": "",
                  "name": "", # lv name
                  "type": ""
                }
              ]
            }
          ]
        },
        "btrfs": {
          "volumes": [ # only one supported, but maybe we can support multiple in the future
            {
              "size": 0,
              "subvolumes": [
                {
                  "name": "",
                  "mountpoint": "",
                  "size": 0
                }
              ]
            }
          ]
        }
      }
    }
  }
}
```

The customization is currently enabled only in Fedora.  I plan to add support and tests in RHEL and CentOS in a follow-up PR.

When the new customizations are used, the base partition table for the image type is ignored.  Instead, a new "custom" partition table is constructed from scratch by iterating through the customizations and creating partitions, volume groups and logical volumes, and btrfs volumes and subvolumes as needed.  Boot, EFI, and root partitions are created automatically to make the partition table valid and usable.  Parameters such as boot mode and partition table type should be determined by the caller of the function, which should be based on the distro, architecture, and image type.  This is, in a way, much simpler than our existing `ensureLVM()` and `ensureBtrfs()` functions, which needed to modify an existing partition table's structure in a way that preserved some parts and not others.

The custom partition table creation function (`NewCustomPartitionTable()`) supports all kinds of valid but perhaps unwanted configurations.  For example, it's possible to create a partition table with both LVM volume groups and btrfs volumes, or multiples of each.  The customizations however don't allow this, but the structure supports it (see above) in case we want to expand what we support in the future.

Customization validation happens in the blueprint.  Invalid values such as duplicate mountpoints, empty btrfs subvolume names, etc are caught by a validator in the blueprint code itself.

What's next:
- Enable the new customizations in RHEL and CentOS.
- Add support for swap partitions and swap logical volumes.
    - I already started on this but removed it from the current PR because it was already getting too long.
- Add support for LUKS encryption.
- **Integrate into osbuild-composer**: I'd like to have this prepared before this PR is merged, which is why I'm keeping this as draft.

mvo5 commented on pkg/disk/disk.go:
```diff
@@ -58,6 +59,54 @@ const (
+	}
+}
+
+func FSTypeFromString(s string) (FSType, error) {
```
(nitpick) maybe just `NewFSType(s) (fstype, error)` ?

achilleas-k replied:
Yeah, `New` is nicer.  I was basing this off the `Arch` enum where we have an `arch.FromString(str)` function, but maybe we can change that too.

mvo5 replied:
Fwiw, it's a nitpick, fine to leave and get back to later (or ignore), especially if there is already a different pattern.

mvo5 commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,289 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	}
+}
+
+func mkESP(size uint64) Partition {
```
(for later) we should make a note that we use this helper in `otk-gen-partition-table` somehow, would be nice to consolidate this (similar to mkBIOSBoot)

mvo5 reviewed (commented):
I had a quick look (and then ran out of time) - this looks super nice so far, two tiny comments and I will continue later.

A meta-question - how will this interplay with the `otk-gen-partition-table` helper? it seems there is a lot of overlap and we will also need to add lvm/lbtrfs/luks there. Will it be unavoidable that we have two code paths that are doing this full the blueprint and for the otk yaml input or is there a way to share code (or do we do that already because it all maps to the same underlying primitives)?. I will try to understand this question myself a bit better and maybe it's too big to answer in detail in a PR in which case we can just do a OOB brainstorm and then put the summary in here.

ondrejbudai commented on pkg/disk/disk.go:
```diff
@@ -58,6 +59,54 @@ const (
+//
+// There should always be one value for each filesystem type supported by
+// osbuild stages (stages/org.osbuild.mkfs.*) and the unset/none value.
+type FSType uint64
```
What about making this a `string` alias, and use descriptive values? Should make debugging a tad bit simpler. Are there downsides (apart from slightly worse performance)?

achilleas-k replied:
There's not much downside other than making it "easy" to use strings in place of the actual value sometimes, which can make it hard to change or track.  Having the enum backed by a uint says, a bit more clearly, "the underlying value doesn't matter", which I find encourages better use of the enum.

But I can go either way.

mvo5 replied:
Fwiw, I had the same thought and tried it out and it was the same about the amount of code/complexity, golang just does not have a good way of doing this it feels so  I did not comment (maybe I should have).

achilleas-k replied:
Do we have any strong feelings either way?  Should I go with strings?

ondrejbudai replied:
I personally don't have strong feelings. :)

ondrejbudai commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,291 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	var bootType FSType
+	switch options.DefaultFSType {
+	case FS_EXT4, FS_XFS:
+		bootType = options.DefaultFSType
```
I suggest documenting in `CustomPartitionTableOptions.DefaultFSType` that the default fs type is also used for `/boot`.

ondrejbudai commented on pkg/blueprint/filesystem_customizations.go:
```diff
@@ -9,9 +9,147 @@ import (
+
+	mountpoints := make(map[string]bool)
+	if p.Plain != nil {
+		for _, mp := range p.Plain.Filesystems {
```
We should also validate types here (and for LVM) I think, so we return an error early on garbage filesystem types.

Interesting cases are btrfs and vfat. Btrfs is definitely a no-go under plain, users should just go through the btrfs field. Btrfs under LVM needs the structure to get expanded. While vfat under plain and LVM is sane, it's extremely rare (apart from /boot/efi ofc), so I would probably error on it until someone actually wants it.

ondrejbudai reviewed (commented):
Awesome work! I love it! I don't see any functional issues, just some minor things that I noticed.

Regarding OTK: The current plan is that osbuild-composer will switch to an OTK backend (through `images`) at some point, right? Then, we will need to somehow pass blueprints to OTK: So either we will have to translate both partitioning customizations to a unified format, or omnifests (and related externals) will simply have to understand both.

achilleas-k commented on pkg/blueprint/filesystem_customizations.go:
```diff
@@ -9,9 +9,60 @@ import (
+}
+
+type LVMCustomization struct {
+	VolumeGroups []VGCustomization `json:"volume-groups,omitempty" toml:"volume-groups,omitempty"`
```
I prefer the "look" of `kebab-case` but `snake_case` plays more nicely with variable names in other contexts (`jq` gets annoying with `-`, `_` is a valid identifier character in most languages whereas `-` is not, etc).

achilleas-k replied:
Let's switch to `_` then.

achilleas-k commented on pkg/blueprint/filesystem_customizations.go:
```diff
@@ -9,9 +9,60 @@ import (
+
+type BtrfsVolumeCustomization struct {
+	// Size of the btrfs partition/volume.
+	MinSize    uint64 `json:"minsize" toml:"minsize"`
```
This is a single volume (partition) that's formatted as btrfs.  The `BtrfsCustomization` theoretically supports multiple volumes/partitions, though we limit it to 1.

ondrejbudai replied:
ofc, I'm blind. Sorry for this comment.

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -413,6 +425,19 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
 
+	requiredSizes := options.RequiredSizes
+	// If no separate requiredSizes are given then we use our defaults
+	if requiredSizes == nil {
```
Yeah, I think you're right.  It's a useful feature, but a built-in default for this is weird.

achilleas-k replied:
Removed from the function internals and added the existing default to the Fedora image types.

mvo5 reviewed (commented):
Some quick comments on the first part, did not manage to get further yet, will try later today.

ondrejbudai commented on pkg/disk/partition_table.go:
```diff
@@ -312,6 +312,55 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
+				}
+				payload.Subvolumes = append(payload.Subvolumes, rootsubvol)
+				break PartitionLoop
+			}
```
What was I thinking? Let's pretend my comment never happened, sorry. :(

achilleas-k commented on pkg/blueprint/filesystem_customizations.go:
```diff
@@ -66,6 +61,97 @@ type BtrfsSubvolumeCustomization struct {
+	mountpoints := make(map[string]bool)
+	if p.Plain != nil {
+		for _, mp := range p.Plain.Filesystems {
+			if mp.Mountpoint == "" {
```
Mountpoint validator added!!

achilleas-k commented:
Fixed up the following from comments:
- Moved required sizes out of the function.  Now `nil` means no requirements.  Added the defaults to all the Fedora image types except the `iot-raw-image` because that had required sizes explicitly disabled.
    - Updated tests.
- All new JSON tags are `snake_case`.
- Mountpoint validation: Previously the validator only checked for empty strings.  Now it also makes sure all mountpoints are absolute paths and "clean" (`filepath.Clean(path) == path`).
    - Extra tests added.
- Replaced all occurrences of `MebiByte` and `GibiByte` in `TestNewCustomPartitionTable()` with `MiB` and `GiB` because it looks nicer.

mvo5 commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,282 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+}
+
+// NewCustomPartitionTable creates a partition table based almost entirely on the partitioning customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
```
(nitpick) This is a long function. It's okay because it is very linear so it's not a big deal, but I was wondering if it might be split up so that parts can be reused for otk and also if it would be easier to follow/extend with splitting (and test individually). I did a quick experiment in  https://github.com/osbuild/images/commit/acb38251b703bb3d4564ad2d2f8769ea864ce226#diff-ffc14a35d85eb6fd9724427b51786e40bc3e910bc41cbd6762ab977ae9689765R455 which I kinda like but maybe just personal preferences - can of course be a followup

mvo5 replied:
Thanks for extracting helpers here, it looks much nicer now IMHO

mvo5 reviewed (commented):
Thank you! This is very nice and will be a very useful feature for our users. I look forward seeing it used (in e.g. bib). I added a bunch of comments/ideas etc but all optional.

What is missing to get it out of draft? I also wonder if might make sense to split into smaller chunks as reviewing 2k lines is not easy but then it's hard to split I guess.

Happy to approve fwiw

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,291 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	// If no separate requiredSizes are given then we use our defaults
+	if requiredSizes == nil {
+		requiredSizes = map[string]uint64{
+			"/":    1073741824,
```
This is gone now and defined on Fedora image types, with GiB.

achilleas-k commented on test/config-map.json:
```diff
@@ -330,5 +330,29 @@
+  },
+  "./configs/partitioning-plain.json": {
+    "image-types": [
+      "ami"
```
AMIs are uploaded to AWS and booted in our testing.

achilleas-k commented on test/configs/partitioning-btrfs.json:
```diff
@@ -0,0 +1,59 @@
+{
+  "name": "partitioning-btrfs",
```
By default, every configuration is built in gitlab CI.  AMIs are also boot tested (push to AWS, boot, SSH in).  No real functional tests are happening yet.
Plans to boot test every image type stalled (we should get on that at some point).

achilleas-k commented on pkg/distro/fedora/imagetype.go:
```diff
@@ -151,6 +151,24 @@ func (t *imageType) getPartitionTable(
 	}
 
 	imageSize := t.Size(options.Size)
+	partitioning := customizations.GetPartitioning()
```
Yup!

achilleas-k commented on pkg/blueprint/filesystem_customizations.go:
```diff
@@ -24,6 +162,24 @@ func (fsc *FilesystemCustomization) UnmarshalTOML(data interface{}) error {
 		return fmt.Errorf("TOML unmarshal: mountpoint must be string, got %v of type %T", d["mountpoint"], d["mountpoint"])
 	}
 
+	switch d["type"].(type) {
```
Adding tests to the existing functions.

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,291 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	switch options.DefaultFSType {
+	case FS_EXT4, FS_XFS:
+		bootType = options.DefaultFSType
+	case FS_NONE:
```
Actually, NONE should be allowed.  I think we should make it so that `options.DefaultFSType == FS_NONE` is only an issue when a non-boot partition is automatically created.

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,291 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+			rootpart := Partition{
+				Type:     FilesystemDataGUID,
+				Bootable: false,
+				Size:     0,
```
I don't know.  0 means autoadjust only in the case of `/` and even then it might stay 0 if there are no required sizes specified.  I think a comment is the right way to go instead of a const that could imply more.

achilleas-k commented on pkg/disk/partition_table_test.go:
```diff
@@ -105,3 +108,632 @@ func TestPartitionTable_GenerateUUIDs_VFAT(t *testing.T) {
+	}
+
+	testCases := map[string]testCase{
+		"null": {
```
Renamed.

achilleas-k commented on pkg/disk/partition_table_test.go:
```diff
@@ -105,3 +108,632 @@ func TestPartitionTable_GenerateUUIDs_VFAT(t *testing.T) {
+			require.NoError(err)
+			require.Equal(tc.expected, pt)
+		})
+	}
```
Agreed.

achilleas-k replied:
This is one of the main reasons I kept this as draft.  I'll add some now and I think we'll be ready to go.

achilleas-k commented on pkg/distro/fedora/imagetype.go:
```diff
@@ -151,6 +151,24 @@ func (t *imageType) getPartitionTable(
+		partOptions := &disk.CustomPartitionTableOptions{
+			PartitionTableType: basePartitionTable.Type, // copy from image type base
+			BootMode:           t.BootMode(),
+			DefaultFSType:      disk.FS_EXT4, // default fs type for Fedora (when not btrfs)
```
> Maybe also worth a small comment that the default fs is not customizable currently (which is curious now that I think about it because the user can de-facto have a default FS by just specifiying all mount points manually and adding their preferred FStype - I forgot if we ever talked about this).

Yes, the user can specify all partitions and their fs type.  There's no way for them to specify what the root filesystem should be if it's automatically added however.  I think as a customization it's kinda pointless: why have a customization for changing the fs type of the automatically created `/` when it's the same as simply specifying `{"mountpoint": "/", "type": "whatever"}`.  So this is a distro-wide policy of "default fs type", which is usually ext4 for Fedora and xfs for RHEL/CentOS.

> Also maybe worth clarifying a bit more what "(when not btrfs)" means, maybe something like `// ext4 is the default on fedora for all non-btrfs partitions, btrfs is modeled as a partition type not a filesystem type here` (or something, I'm not happy with my comment, it feels too long).

eh, I'd rather just remove the comment or keep "ext4 is the default for Fedora".  The "when not btrfs" was more of a clarification for the mechanics of automatic subvolumes being created on btrfs, but I think that doesn't need clarification tbh.

achilleas-k commented on pkg/distro/fedora/imagetype.go:
```diff
@@ -151,6 +151,24 @@ func (t *imageType) getPartitionTable(
 	imageSize := t.Size(options.Size)
+	partitioning := customizations.GetPartitioning()
+	if partitioning != nil {
+		// Use the new custom partition table to create a PT fully based on the user's customizations.
```
This is checked in `checkOptions()`.  We could do it at the blueprint level but I'd rather remove the old filesystem customizations from the bp code in images soon after this is merged anyway.

achilleas-k commented on pkg/distro/fedora/imagetype.go:
```diff
@@ -151,6 +151,24 @@ func (t *imageType) getPartitionTable(
+		}
+
+		partOptions := &disk.CustomPartitionTableOptions{
+			PartitionTableType: basePartitionTable.Type, // copy from image type base
```
Done!

achilleas-k commented on pkg/distro/fedora/distro_test.go:
```diff
@@ -937,6 +937,51 @@ func TestDistro_CustomUsrPartitionNotLargeEnough(t *testing.T) {
+			for _, imgTypeName := range arch.ListImageTypes() {
+				imgType, _ := arch.GetImageType(imgTypeName)
+				_, _, err := imgType.Manifest(&bp, distro.ImageOptions{}, nil, 0)
+				if imgTypeName == "iot-commit" || imgTypeName == "iot-container" || imgTypeName == "iot-bootable-container" {
```
Let's rewrite these separately with cases since there's a lot of tests like this all over the place.

achilleas-k commented:
I think I covered all the existing comments and I expanded the test matrix for the new function.

Marking as ready.

ondrejbudai reviewed (dismissed):
There's a tiny documentation issue, but we can resolve it in a follow-up. Otherwise, this LGTM. Thanks a ton! :heart:

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,144 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	// mountpoint that contains the directory. Additional directory
+	// requirements are additive, meaning the minimum size for a mountpoint's
+	// partition is the sum of all the required directory sizes it will
+	// contain. By default, if no requiredSizes are provided, the new partition
```
Oops!  Thanks for catching that.

achilleas-k replied:
Fixed!

ondrejbudai reviewed (approved):
You are fast! :heart_eyes:

ondrejbudai reviewed (changes requested):
Lemme go over user feedback that I just got before merging it. :)

bcl commented on pkg/disk/partition_table.go:
```diff
@@ -188,6 +188,9 @@ func mkESP(size uint64) Partition {
 
 type CustomPartitionTableOptions struct {
+	// PartitionTableType must be either "dos" or "gpt". Defaults to "gpt".
+	PartitionTableType string
```
We have an enum-like type for filesystem types, I think it would make sense to be consistent and do the same here instead of strings.

achilleas-k replied:
Yeah, good idea.

achilleas-k replied:
Done.

bcl commented on pkg/disk/partition_table.go:
```diff
@@ -204,6 +204,15 @@ type CustomPartitionTableOptions struct {
+	// requirements are additive, meaning the minimum size for a mountpoint's
+	// partition is the sum of all the required directory sizes it will
+	// contain.
+	RequiredSizes map[string]uint64
```
nitpick: I'd call it `RequiredMinSizes` or maybe reuse `FilesystemCustomization` so that it's clear just from the variable name that these are minimums and not an exact size (even though the docstring clearly explains that).

achilleas-k replied:
`RequiredMinSizes` sounds good.  I think originally "required" was meant to imply "lower bound"/"minimum requirements" sort of, but adding `Min` to the mix makes it clearer.

achilleas-k replied:
Renamed.

bcl commented:
Wow, great job with this! I only had a couple small comments and one thing I didn't see -- is there anything that catches overrlapping partitions? I didn't notice anything in the validation function or tests.

achilleas-k commented:
> Wow, great job with this! I only had a couple small comments and one thing I didn't see -- is there anything that catches overrlapping partitions? I didn't notice anything in the validation function or tests.

The validator in the blueprint normalises mountpoints and checks for duplicates across the whole customization block.  It's probably a good idea to have the same checks in the `NewCustomPartitionTable()` function since it can be used externally without going through the blueprint (like, if BIB consumes the function directly somehow).

achilleas-k commented:
> > Wow, great job with this! I only had a couple small comments and one thing I didn't see -- is there anything that catches overrlapping partitions? I didn't notice anything in the validation function or tests.
> 
> The validator in the blueprint normalises mountpoints and checks for duplicates across the whole customization block. It's probably a good idea to have the same checks in the `NewCustomPartitionTable()` function since it can be used externally without going through the blueprint (like, if BIB consumes the function directly somehow).

See last commit: `blueprint: split PartitioningCustomization validation function`

achilleas-k commented:
Rebased and resolved conflicts.

mvo5 commented on pkg/disk/disk.go:
```diff
@@ -61,6 +62,89 @@ const (
+//
+// There should always be one value for each filesystem type supported by
+// osbuild stages (stages/org.osbuild.mkfs.*) and the unset/none value.
+type FSType uint64
```
I feel `FSType` is short but clear (unlike `PTType` see my comment below)

mvo5 commented on pkg/blueprint/filesystem_customizations.go:
```diff
@@ -86,3 +260,45 @@ func CheckMountpointsPolicy(mountpoints []FilesystemCustomization, mountpointAll
+	}
+
+	// collect all mountpoints
+	mountpoints := make([]string, 0)
```
The one tiny advantage is that `var mountpoints []string` does not require a function call and mountpoints is just `nil` whereas `make` createa a empty slice.

mvo5 commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,146 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+
+	// validate the partitioning customizations before using them
+	if err := customizations.Validate(); err != nil {
+		return nil, fmt.Errorf("error generating partition table: %w", err)
```
(nitpick) maybe worth extracting a common `errPrefix` here and just reuse that? Something like:
```go
diff --git a/pkg/disk/partition_table.go b/pkg/disk/partition_table.go
index 8a682668e..a4361730b 100644
--- a/pkg/disk/partition_table.go
+++ b/pkg/disk/partition_table.go
@@ -242,9 +242,10 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
                customizations = &blueprint.PartitioningCustomization{}
        }
 
+       errPrefix := "error generating partition table: "
        // validate the partitioning customizations before using them
        if err := customizations.Validate(); err != nil {
-               return nil, fmt.Errorf("error generating partition table: %w", err)
+               return nil, fmt.Errorf(errPrefix+"%w", err)
        }
 
        pt := &PartitionTable{}
@@ -256,7 +257,7 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
                // default to "gpt"
                pt.Type = PT_GPT.String()
        default:
-               return nil, fmt.Errorf("error generating partition table: invalid partition table type enum value: %d", options.PartitionTableType)
+               return nil, fmt.Errorf(errPrefix+"invalid partition table type enum value: %d", options.PartitionTableType)
        }
 
        // The boot type will be the default only if it's a supported filesystem
@@ -271,7 +272,7 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
        }
 
        if err := addPartitionsForBootMode(pt, options.BootMode); err != nil {
-               return nil, fmt.Errorf("error generating partition table: %w", err)
+               return nil, fmt.Errorf(errPrefix+"%w", err)
        }
        if customizations.LVM != nil || customizations.Btrfs != nil {
                // we need a /boot partition to boot LVM or Btrfs, create boot
@@ -280,10 +281,10 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
        }
 
        if err := addPartitionsForPlain(pt, customizations.Plain, options); err != nil {
-               return nil, fmt.Errorf("error generating partition table: %w", err)
+               return nil, fmt.Errorf(errPrefix+"%w", err)
        }
        if err := addPartitionsForLVM(pt, customizations.LVM, options); err != nil {
-               return nil, fmt.Errorf("error generating partition table: %w", err)
+               return nil, fmt.Errorf(errPrefix+"%w", err)
        }
        addPartitionsForBtrfs(pt, customizations.Btrfs)
 

```

achilleas-k replied:
Done, but I used `("%s %w", errPrefix, err)`.

mvo5 commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,146 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	if customizations.LVM != nil || customizations.Btrfs != nil {
+		// we need a /boot partition to boot LVM or Btrfs, create boot
+		// partition if it does not already exist
+		ensureBoot(pt, bootType)
```
(super nitpick) the name `ensureBoot` breaks the pattern of `addPartitionsFor*` that the we here maybe something `ensureBootPartition()` or `addBootPartition()` to be slightly more similar to the other names?

achilleas-k replied:
`ensureBootPartition()` works.  I wouldn't want to call it `addBootPartition()` though.  I think using `ensure` in these functions (for root too) is a good way to imply that it might not create anything (if it already exists), whereas the `add...` functions always create things (assuming the args aren't nil).

achilleas-k replied:
Renamed.

mvo5 commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,146 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	// The boot type will be the default only if it's a supported filesystem
+	// type for /boot (ext4 or xfs). Otherwise, we default to xfs.
+	// FS_NONE also falls back to xfs.
+	var bootType FSType
```
(super nitpick) maybe `bootFSType` or `bootFstype` to make it more distinct from `options.BootMode` ?

achilleas-k replied:
Done

mvo5 commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,146 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	// type for /boot (ext4 or xfs). Otherwise, we default to xfs.
+	// FS_NONE also falls back to xfs.
+	var bootType FSType
+	switch options.DefaultFSType {
```
(nitpick) maybe move this closer to line 276 where it is used? Right now it looks like it's related to `addPartitionsForBootMode()` but for that it's not actually needed (i.e. something like):
```go
diff --git a/pkg/disk/partition_table.go b/pkg/disk/partition_table.go
index 8a682668e..31de9f783 100644
--- a/pkg/disk/partition_table.go
+++ b/pkg/disk/partition_table.go
@@ -262,18 +262,19 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
        // The boot type will be the default only if it's a supported filesystem
        // type for /boot (ext4 or xfs). Otherwise, we default to xfs.
        // FS_NONE also falls back to xfs.
-       var bootType FSType
-       switch options.DefaultFSType {
-       case FS_EXT4, FS_XFS:
-               bootType = options.DefaultFSType
-       default:
-               bootType = FS_XFS
-       }
 
        if err := addPartitionsForBootMode(pt, options.BootMode); err != nil {
                return nil, fmt.Errorf("error generating partition table: %w", err)
        }
+
        if customizations.LVM != nil || customizations.Btrfs != nil {
+               var bootType FSType
+               switch options.DefaultFSType {
+               case FS_EXT4, FS_XFS:
+                       bootType = options.DefaultFSType
+               default:
+                       bootType = FS_XFS
+               }
                // we need a /boot partition to boot LVM or Btrfs, create boot
                // partition if it does not already exist
                ensureBoot(pt, bootType)

```

achilleas-k replied:
Done

mvo5 commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,146 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+		return nil, err
+	}
+
+	if requiredSizes := options.RequiredMinSizes; len(requiredSizes) != 0 {
```
(super nitpick) not sure we need the temporary assignment here (or am I missing something?), maybe just:
```go
diff --git a/pkg/disk/partition_table.go b/pkg/disk/partition_table.go
index 8a682668e..36de2f707 100644
--- a/pkg/disk/partition_table.go
+++ b/pkg/disk/partition_table.go
@@ -291,8 +292,8 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
                return nil, err
        }
 
-       if requiredSizes := options.RequiredMinSizes; len(requiredSizes) != 0 {
-               pt.EnsureDirectorySizes(requiredSizes)
+       if len(options.RequiredMinSizes) != 0 {
+               pt.EnsureDirectorySizes(options.RequiredMinSizes)
        }
 
        pt.relayout(customizations.MinSize)
```

achilleas-k replied:
Done.

mvo5 reviewed (commented):
Some initial quick comment, mostly tiny nitpicks/ideas/suggestions :)

achilleas-k commented on pkg/disk/disk.go:
```diff
@@ -61,6 +62,89 @@ const (
+}
+
+// PTType is the partition table type enum.
+type PTType uint64
```
`types` might be a bit much and it's too generic.  It'd be nice if we could have something short that said "partition table and disk related types" instead of implying a generic package for types.

mvo5 commented on pkg/blueprint/filesystem_customizations.go:
```diff
@@ -4,14 +4,227 @@ import (
 	"github.com/osbuild/images/pkg/pathpolicy"
 )
 
+type PartitioningCustomization struct {
```
Maybe put this into a new `partition_customizations.go` ? As they are distinict from the tranditional mountpoint based filesystem customizations?

achilleas-k replied:
Done.

mvo5 reviewed (commented):
Thanks, this looks very nice, it's also very long which makes it hard to review. I did another pass, at the end my attention span wasn't great anymore so I will probably have another look *or* we could merge it and followup, I am fine with the later but there are one or two questions (e.g. the one about potentially duplicated subvolume names) that I would like to double check on first before setting this to approved.

achilleas-k commented on pkg/blueprint/partitioning_customizations.go:
```diff
@@ -0,0 +1,261 @@
+// defined. Returns an error if both LVM and btrfs are set and if either has
+// more than one element.
+// This function also calls [PartitioningCustomization.Validate].
+func (p *PartitioningCustomization) ValidateSimple() error {
```
This is 100% a naming issue.  Validate simple sounds like it does a simpler form of validation, but instead it's meant to imply that it's requiring a simpler set of customizations (no mixing btrfs with lvm for example).  We need a better name for it.  Something that says "Validate but apply the stricter rules as well" (1 volume, no mixing).

mvo5 replied:
Right, that makes sense from an implementation point of view. I wonder if from an API/consumer point of view it still wouldn't make sense to flip it around, i.e. as a user of this API I want to call `PartitioningCustomization.Validate()` - as an (API) user the fact that we (somewhat artificially) limit what can be used does not matter, all I want is a (single) validate (I think).

So maybe something like: 
```go
diff --git a/pkg/blueprint/partitioning_customizations.go b/pkg/blueprint/partitioning_customizations.go
index 280f745bf..da1b260df 100644
--- a/pkg/blueprint/partitioning_customizations.go
+++ b/pkg/blueprint/partitioning_customizations.go
@@ -92,11 +92,10 @@ func validateFilesystemType(path, fstype string) error {
        return nil
 }
 
-// ValidateSimple checks that at most one LVM Volume Group or btrfs volume is
-// defined. Returns an error if both LVM and btrfs are set and if either has
-// more than one element.
-// This function also calls [PartitioningCustomization.Validate].
-func (p *PartitioningCustomization) ValidateSimple() error {
+// validateLayoutConstraints checks that at most one LVM Volume Group
+// or btrfs volume is defined. Returns an error if both LVM and btrfs
+// are set and if either has more than one element.
+func (p *PartitioningCustomization) validateLayoutConstraints() error {
        if p == nil {
                return nil
        }
@@ -113,7 +112,7 @@ func (p *PartitioningCustomization) ValidateSimple() error {
                return fmt.Errorf("multiple LVM volume groups are not yet supported")
        }
 
-       return p.Validate()
+       return nil
 }
 
 // Validate checks for customization combinations that are generally not
@@ -124,6 +123,10 @@ func (p *PartitioningCustomization) Validate() error {
                return nil
        }
 
+       if err := p.validateLayoutConstraints(); err != nil {
+               return err
+       }
+
        // iterate through everything and look for:
        // - invalid mountpoints (global)
        // - duplicate mountpoints (global)
diff --git a/pkg/blueprint/partitioning_customizations_test.go b/pkg/blueprint/partitioning_customizations_test.go
index d469acc6a..de61be42a 100644
--- a/pkg/blueprint/partitioning_customizations_test.go
+++ b/pkg/blueprint/partitioning_customizations_test.go
@@ -660,7 +660,7 @@ func TestPartitioningValidation(t *testing.T) {
        for name := range testCases {
                tc := testCases[name]
                t.Run(name, func(t *testing.T) {
-                       err := tc.partitioning.ValidateSimple()
+                       err := tc.partitioning.Validate()
                        if tc.expectedMsg != "" {
                                assert.EqualError(t, err, tc.expectedMsg)
                        } else {
```
(names TBD) - but maybe I'm just missing something here?

achilleas-k replied:
The problem I was trying to solve is being able to validate without the layout constraints, so that the blueprint code itself applies the extra constraints (at the UI level) but the internals (`NewCustomPartitionTable()`) only check for invalid configurations like conflicting names and labels.

I think I'm going to take your suggestion but keep them separate, make `validateLayoutConstraints()` private like you suggest and run it completely separately.  So neither function would call the other.

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -160,6 +161,148 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+// NewCustomPartitionTable creates a partition table based almost entirely on the partitioning customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
+	if options == nil {
+		// init options with empty struct to use defaults
```
Yeah, I think the comment was written when it was still empty.

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -715,7 +858,7 @@ func (pt *PartitionTable) ensureLVM() error {
 		// create root logical volume on the new volume group with the same
 		// size and filesystem as the previous root partition
-		_, err := vg.CreateLogicalVolume("root", part.Size, filesystem)
+		_, err := vg.CreateLogicalVolume("rootlv", part.Size, filesystem)
```
This commit doesn't affect the default image definitions, but it does change the behaviour of `vg.CreateLogicalVolume()`.  It also doesn't affect the behaviour of `CreateMountpoint()`.

I had to make this change so that I can use `CreateLogicalVolume()` without having the lv name change internally in the function.  The function is exported, but it's not used in osbuild-composer directly, so it shouldn't have any effect on the behaviour there.

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -894,3 +1037,207 @@ func (pt *PartitionTable) GetMountpointSize(mountpoint string) (uint64, error) {
+			return nil
+		case *Btrfs:
+			rootsubvol := BtrfsSubvolume{
+				Name:       "root",
```
Good catch.  Yes, that would be a problem.

achilleas-k commented on pkg/disk/disk.go:
```diff
@@ -61,6 +62,89 @@ const (
+	}
+}
+
+func PTTypeFromString(s string) (PartitionTableType, error) {
```
Changing it to `NewPartitionTableType()`.  I'll also change `FSTypeFromString()` to `NewFSType()`.

achilleas-k commented on pkg/disk/partition_table.go:
```diff
@@ -894,3 +1037,207 @@ func (pt *PartitionTable) GetMountpointSize(mountpoint string) (uint64, error) {
+		// add both
+		pt.Partitions = append(pt.Partitions, mkBIOSBoot())
+		pt.Partitions = append(pt.Partitions, mkESP(200*datasizes.MiB))
+	case platform.BOOT_NONE:
```
To remove the `return nil` at the end we need to add returns to each case.
Which is fair enough, no strong feelings either way.

achilleas-k commented on pkg/blueprint/partitioning_customizations_test.go:
```diff
@@ -0,0 +1,671 @@
+		t.Run(name, func(t *testing.T) {
+			err := tc.partitioning.ValidateSimple()
+			if tc.expectedMsg != "" {
+				assert.EqualError(t, err, tc.expectedMsg)
```
Not sure the latter is better or worse.  If the expected message is long, it might be a distraction when the more important information is that an error wasn't returned, which you get in the `Error:` field.  When they don't match, you get both the expected and the actual in the `Error:` field, so repeating the expected message is extra noise (for long messages).

achilleas-k commented:
I decided (read, I was kindly asked) to split this PR into more manageable chunks, which I totally agree is a good move.

Closing this now.
